File: src/dotnet/clrfuncreflectionwrap.cpp
#include "edge.h"

ClrFuncReflectionWrap::ClrFuncReflectionWrap()
{
    // empty
}

ClrFuncReflectionWrap^ ClrFuncReflectionWrap::Create(
    Assembly^ assembly, System::String^ typeName, System::String^ methodName)
{
    System::Type^ startupType = assembly->GetType(typeName, true, true);
    ClrFuncReflectionWrap^ wrap = gcnew ClrFuncReflectionWrap();
    wrap->instance = System::Activator::CreateInstance(startupType, false);
    wrap->invokeMethod = startupType->GetMethod(methodName, BindingFlags::Instance | BindingFlags::Public);
    if (wrap->invokeMethod == nullptr) 
    {
        throw gcnew System::InvalidOperationException(
            "Unable to access the CLR method to wrap through reflection. Make sure it is a public instance method.");
    }
    
    return wrap;
}

Task<System::Object^>^ ClrFuncReflectionWrap::Call(System::Object^ payload)
{
    return (Task<System::Object^>^)this->invokeMethod->Invoke(
                this->instance, gcnew array<System::Object^> { payload });
}

File: src/common/clrfuncreflectionwrap.cs
using System;
using System.Threading.Tasks;
using System.Reflection;

public class ClrFuncReflectionWrap
{
    Object instance;
    MethodInfo invokeMethod;

    public static ClrFuncReflectionWrap Create(Assembly assembly, String typeName, String methodName)
    {
        Type startupType = assembly.GetType(typeName);

        if (startupType == null)
        {
            throw new TypeLoadException("Could not load type '" + typeName + "'");
        }

        ClrFuncReflectionWrap wrap = new ClrFuncReflectionWrap();
        wrap.instance = System.Activator.CreateInstance(startupType);
        wrap.invokeMethod = startupType.GetMethod(methodName, BindingFlags.Instance | BindingFlags.Public);
        if (wrap.invokeMethod == null)
        {
            throw new System.InvalidOperationException(
                "Unable to access the CLR method to wrap through reflection. Make sure it is a public instance method.");
        }

        return wrap;
    }

    public Task<Object> Call(Object payload)
    {
        return (Task<Object>)this.invokeMethod.Invoke(this.instance, new object[] { payload });
    }
};
